Core/Src/main.c

```diff
@@ -1,181 +1,215 @@
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+	// To access Channel 0 of ADC MCP3004
+	uint8_t transmit_data[] = {0x01, 0x80, 0x00};
+
+	// store receive bits of DV from ADC
+	uint8_t receive_data[3] = {};
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+
+  while (1)
+  {
+	  // STEP 1: Coding ADC Communication
+	  //CS of SPI should be low when sampling the analog voltage
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+
+	  HAL_SPI_TransmitReceive(&hspi1, transmit_data, receive_data, sizeof(receive_data) ,100);
+
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+	  // STEP 2: Converting ADC value to PWM Signal
+
+	  // Access the 10bit DV from receive_data
+	  uint8_t rd_1 = receive_data[1];
+	  uint16_t DV = (rd_1 & 0x03) << 8 | receive_data[2];
+
+
+	  // 1)Change the 10 bit DV we get from the ADC to the number of counts of On time
+	  uint16_t duty_cycle = (DV / 1023) + 1;

```

Comment:
the 1023 here appears as a magic number. I knon it means 2^10 -1, but other people might not get the math here. The recommanded way of doing this is to make a macro such as #define MAX_ADC 1023, and use MAX_ADC in our calculation

Reply:
Additionally, if I understand correctly, the value of duty_cycle here is 1+ percentage of adc. But remember, we want it to vary from 5% to 10% of the duty cycle. This is gonna be 5% to 10% timer counts in a timer period. Your timer period currently is 65535 and 5% of it should be  3277 ticks. The number of ticks is what you should sent to the compare function.